src/java/org/apache/cassandra/db/ReadCommand.java

```diff
@@ -894,6 +896,134 @@ protected LongPredicate getPurgeEvaluator()
+     * Other (non-purgeable) tombstones will be tracked by regular Cassandra logic later.
+     */
+    private UnfilteredPartitionIterator withPurgeableTombstonesMetricRecording(UnfilteredPartitionIterator iter,
+                                                                               ColumnFamilyStore cfs)
+    {
+        class PurgeableTombstonesMetricRecording extends Transformation<UnfilteredRowIterator>
+        {
+            private int purgeableTombstones = 0;
+
+            @Override
+            public UnfilteredRowIterator applyToPartition(UnfilteredRowIterator iter)
+            {
+                if (!iter.partitionLevelDeletion().isLive())
+                    purgeableTombstones++;
+                return Transformation.apply(iter, this);
+            }
+
+            @Override
+            public Row applyToStatic(Row row)
+            {
+                return applyToRow(row);
+            }
+
+            @Override
+            public Row applyToRow(Row row)
+            {
+                final long nowInSec = nowInSec();
+                boolean hasTombstones = false;
+
+                if (DatabaseDescriptor.getPurgeableTobmstonesMetricGranularity() == Config.TombstonesMetricGranularity.cell)
+                    for (Cell<?> cell : row.cells())
+                    {
+                        if (!cell.isLive(nowInSec) && isPurgeable(cell.localDeletionTime(), nowInSec))
+                        {
+                            purgeableTombstones++;
+                            hasTombstones = true; // allows to avoid counting an extra tombstone if the whole row expired
+                        }
+                    }
+
+                // we replicate the logic is used for non-purged tombstones metric here
+                if (!row.primaryKeyLivenessInfo().isLive(nowInSec)
+                    && row.hasDeletion(nowInSec)
+                    && isPurgeable(row.deletion().time(), nowInSec)
+                    && !hasTombstones)
+                {
+                    // We're counting primary key deletions only here.
+                    purgeableTombstones++;
+                }
+
+                return row;
+            }
+
+            @Override
+            public RangeTombstoneMarker applyToMarker(RangeTombstoneMarker marker)
+            {
+                final long nowInSec = nowInSec();
+
+                // for boundary markers - increment metric only if both - close and open - markers are purgeable
+                if (marker.isBoundary())
+                {
+                    countIfBothPurgeable(marker.closeDeletionTime(false),
+                                         marker.openDeletionTime(false),
+                                         nowInSec);
+                }
+                // for bound markers - just increment if it is purgeable
+                else
+                {
+                    countIfPurgeable(((RangeTombstoneBoundMarker) marker).deletionTime(), nowInSec);
```

Comment:
Bold cast! How do you know it is always like that?

Reply:
@netudima could we at least check that `marker` is instance of `RangeTombstoneBoundMarker` before casting? If it is not instance of `RangeTombstoneBoundMarker`, we should just skip this logic.

Reply:
Yes, I am planning to do this, just have not committed yet

Reply:
done